bindings/python/CONTRIBUTING.md

```diff
@@ -0,0 +1,790 @@
+  or working with the object use `.get()` function which returns the pointer.
+
+```cython
+from libcpp.memory cimport shared_ptr
+lhs.thisptr.get().add_state()
+```
+
+## Supporting Python for loops
+
+* **Problem**: You want iterate over your structure in Python `for` loops.
+* **Solution**: The standard function should support this by default (i.e., `std::` containers),
+  your custom classes needs to support basic iterators (`begin()`, `end()`, etc.).
+
+
+```cython
+cdef vector[CTrans] c_transitions = self.thisptr.get().get_transitions_to(state_to)
+trans = []
+for c_transition in c_transitions:
+#   ^-- classic for each loop over std::vector<Trans> does everything behind the scenes
+    trans.append(Trans(c_transition.src, c_transition.symb, c_transition.tgt))
+```
+
+## Working with strings
+
+* **Problem**: You want to work with strings and pass them between binding and wrapped library.
+* **Solution**: You need to use `encode()` when passing strings to C/C++ functions, and `decode()`,
+  when recieving results.
+
+```cython
+k.encode('utf-8')
+# ^- when passing strings to C/C++ library you need to encode
+result.decode('utf-8')
+# ^- when retrieving results from C/C++ you need to decode
+```
+
+## Helping Cython to infer types
+
+* **Problem**: Cython fails to infer types for the variables and cannot compile the binding.
+* **Solution**: You need to help Cython a little: add types to parameters, add types to cdefined 
+  variables, or manually retype the variable.
+
+```cython
+(<mata_nfa.Nfa>noodle_segment).thisptr = c_noodle_segment
+# ^-- tell Cython that this variable is of given type
+objects: List[subprocess.Popen] = []
+# ^-- tell Cython the type of the Python object
+cdef mata_nfa.stringstream* output_stream
+# ^-- define variable as C/C++ type
+def union(cls, Nfa lhs, Nfa rhs):
+    pass
+#              ^-- tell Cython types of parameters
+```
+
+## Using pre- and post-increments (`var++` and `++var`)
+
+* **Problem**: You want to manually use pre/post increments in your code.
+* **Solution**: You have to import  `postincrement` and `preincrement` from `cython.operator`.
+
+```cython
+from cython.operator import postincrement as postinc, preincrement as preinc
+postinc(it)
+preinc(iterator)
+```
+
+## Working with exceptions
+
+* **Problem**: Your function can fire exceptions.
+* **Solution**: You have to tell Cython how to handle these.
+
+```cython
+void add(CTrans) except -1
+#                ^-- exceptions are returned as -1 in Cython
+void add(CTrans) except +
+#                ^-- function may fire exception; C/C++ exceptions will be translated to Python exceptions
+void add(CTrans) except +MemoryError
+#                ^-- function may fire MemoryError (corresponds to bad_alloc) 
+void add(CTrans) noexcept +
+#                ^-- function will not fire exception
+```
+
+* Note, there are other ways of handing exceptions, but these should suffice.
+
+# Importing between modules
```

Comment:
Maybe start the whole guide with this section?

Reply:
On second thought, you explain a bit of importing at the start. This should be sufficient. I will close this conversation.